### PR TITLE
Riedl

### PR DESCRIPTION
**Pull request by jriedl**

### CODE REVIEW

File: cpp/new_journal_alert.cc [hunk 1]
```diff
@@ -418,12 +420,56 @@ void LoadBundleControlNumbers(const IniFile &bundles_config, const std::string &
...
+
+
+bool IsBundle(const std::string &serial_control_number) {
+    if (serial_control_number.empty())
+        return false;
+    return not std::isdigit(serial_control_number[0]);
+}
+
+
+bool MinModificationTimeComparator(const std::pair<std::string, std::string> elem1, std::pair<std::string, std::string> elem2) {
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:449]: A better name please like ...LessThan..>?  Also, the second argument should be const, too.

File: cpp/new_journal_alert.cc [hunk 2]
```diff
@@ -418,12 +420,56 @@ void LoadBundleControlNumbers(const IniFile &bundles_config, const std::string &
...
+    return TimeUtil::Iso8601StringToTimeT(elem1.second, TimeUtil::UTC /* required by function */) <
+           TimeUtil::Iso8601StringToTimeT(elem2.second, TimeUtil::UTC /* required by function */);
+}
+
+
+std::string GetMinLastModificationTime(const std::map<std::string, std::string> &control_number_and_max_last_modification_times)
+{
+    auto min(std::min_element(control_number_and_max_last_modification_times.begin(),
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:457]: Missing "const".

File: cpp/new_journal_alert.cc [hunk 3]
```diff
@@ -418,12 +420,56 @@ void LoadBundleControlNumbers(const IniFile &bundles_config, const std::string &
...
+}
+
+
 void ProcessSingleUser(
     const bool debug, DbConnection * const db_connection, const std::unique_ptr<KeyValueDB> &notified_db,
     const IniFile &bundles_config, std::unordered_set<std::string> * const new_notification_ids,
     const std::string &user_id, const std::string &solr_host_and_port, const std::string &hostname,
     const std::string &sender_email, const std::string &email_subject,
-    std::vector<SerialControlNumberAndMaxLastModificationTime> &control_numbers_or_bundle_names_and_last_modification_times)
+    std::vector<SerialControlNumberAndMaxLastModificationTime> &control_numbers_or_bundle_names_and_last_modification_times,
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:471]: Reference arguments should typically be const.

File: cpp/new_journal_alert.cc [hunk 4]
```diff
@@ -501,6 +555,21 @@ void ProcessSingleUser(
 }
 
 
+void StoreBundleJournalsMaxModificationTimes(DbConnection * const db_connection,
+                                              const std::map<std::string, std::map<std::string, std::string>>
+                                              bundle_journals_last_modification_times)
+{
+    for (const auto &bundle : bundle_journals_last_modification_times) {
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:562]: Please use the C++17 syntax [first, second] w/ appropriate names.

File: cpp/new_journal_alert.cc [hunk 5]
```diff
@@ -131,6 +132,33 @@ bool NewIssueInfo::operator<(const NewIssueInfo &rhs) const {
 }
 
 
+bool NewIssueInfo::operator==(const NewIssueInfo &rhs) const {
+    return control_number_ == rhs.control_number_ and
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:136]: Wouldn't this line suffice?

File: cpp/new_journal_alert.cc [hunk 6]
```diff
@@ -131,6 +132,33 @@ bool NewIssueInfo::operator<(const NewIssueInfo &rhs) const {
...
+           issue_title_ == rhs.issue_title_ and
+           volume_ == rhs.volume_ and
+           issue_ == rhs.issue_ and
+           start_page_ == rhs.start_page_ and
+           authors_ == rhs.authors_;
+}
+
+
+} //unnamed namespace
+
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:147]: Missing line.

File: cpp/new_journal_alert.cc [hunk 7]
```diff
@@ -420,7 +448,8 @@ void LoadBundleControlNumbers(const IniFile &bundles_config, const std::string &
 }
 
 
-std::function<std::string(std::string)> SingleQuote = [](std::string to_quote) { return "'" + to_quote + "'"; };
+std::string SingleQuote(std::string to_quote) { return "'" + to_quote + "'"; };
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:451]: Argument should be a const reference.

File: cpp/new_journal_alert.cc [hunk 8]
```diff
@@ -559,11 +597,11 @@ void StoreBundleJournalsMaxModificationTimes(DbConnection * const db_connection,
                                               const std::map<std::string, std::map<std::string, std::string>>
                                               bundle_journals_last_modification_times)
 {
-    for (const auto &bundle : bundle_journals_last_modification_times) {
-        db_connection->queryOrDie("DELETE FROM ixtheo_journal_bundles WHERE bundle_name=\"" + bundle.first + "\"");
-        for (const auto &journal_control_number_and_max_last_modification_time : bundle.second)
+    for (auto const& [bundle_name, journal_control_number_and_max_last_modification_times] : bundle_journals_last_modification_times) {
```
Comment (ruschein) [inline cpp/new_journal_alert.cc:600]: The ampersand goes on the other side just like "int *" and not "int* " etc.

File: cpp/new_journal_alert.cc (hunk 5 above)
Comment (jriedl) [reply cpp/new_journal_alert.cc:136]: Probably mostly yes but might lead to sublte errors.

**Pull request merged**